views/CurrencyConverterSliderItem.ts

```diff
@@ -0,0 +1,161 @@
+type Listener<T> = (ev: T) => void;
+
+type Observer<T> = {
+  subscribe: (listener: Listener<T>) => () => void;
+  publish: (event: T) => void;
+};
+
+function createObserver<T>(): Observer<T> {
+  let listeners: Listener<T>[] = [];
+
+  return {
+    subscribe: (listener: Listener<T>): (() => void) => {
+      listeners.push(listener);
+      return () => {
+        listeners = listeners.filter((l) => l !== listener);
+      };
+    },
+    publish: (event: T) => {
+      listeners.forEach((l) => l(event));
+    },
+  };
+}
+
+type CurrencyChangeEvent = {
+  type: 'EUR' | 'X';
+  value: number;
+};
+
+export interface CurrencyConverterItemSliderProps {
+  currencyName: string;
+  currencyRate: number;
+}
+
+export class CurrencyConverterSliderItem {
+  private _parent: Element;
+  private _euroQuantity: number = 100;
+  private _xName: string = '';
+  private _xRate: number = 0;
+  private _observer: Observer<CurrencyChangeEvent> = createObserver<CurrencyChangeEvent>();
+
+  constructor(root: Element, data: CurrencyConverterItemSliderProps) {
+    this._parent = root;
+
+    this._xName = data.currencyName;
+    this._xRate = data.currencyRate;
+
+    this._observer.subscribe((data) => {
+      const { type, value } = data;
+
+      if (type === 'EUR') {
+        const xCurrency = this.getElement(`.slider-${this.xName}-euro`) as HTMLInputElement;
+        xCurrency.value = parseFloat(`${value * this.xRate}`).toFixed(3);
+      } else {
+        const euro = this.getElement(`.slider-euro-${this.xName}`) as HTMLInputElement;
+        euro.value = parseFloat(`${value / this.xRate}`).toFixed(2);
+      }
+    });
+  }
+
+  getElement(selector: string) {
+    return document.querySelector(selector);
+  }
+
+  getAllElement(selector: string) {
+    return document.querySelectorAll(selector);
+  }
+
+  eventsMap(): { [k: string]: (event) => void } {
+    return {
+      [`change:.slider-euro-${this.xName}`]: this.onEuroCurrencyChangeHandler,
+      [`change:.slider-${this.xName}-euro`]: this.onXCurrencyChangeHandler,
```

Comment:
```this.onEuroCurrencyChangeHandler``` and ```this.onXCurrencyChangeHandler```
these both will be always equal to ```undefined``` since ```this``` refer to the object you return

Reply:
Completely new PR created: https://github.com/CuongNgMan/pattern-of-patterns/pull/2